File: CSharpCodeBase/components/character/vision.lua
local Class = require 'utils.hump.class'
local Component = require 'components.component'

local VisionComponent = Class({__includes=Component, init = function (self, entity)
      Component.init(self, entity)
      self.entity.vision = self
      self.radius = self.entity.config:Get("visionRadius")
      self._tags = {}
end})

function VisionComponent:IsTargetVisible(targetEntity)
  return targetEntity:GetEffectiveDistance(self.entity) < self.radius
end

function VisionComponent:GetVisibleEntities(radius)
  return RaycastUtils.GetEntitiesInRadius(self.entity:GetPosition(), radius or self.radius)
end

function VisionComponent:GetVisibleEntitiesByTypes(types, radius)
  return RaycastUtils.GetEntitiesInRadiusByTypes(self.entity:GetPosition(),
    radius or self.radius,
    types)
end

function VisionComponent:GetVisibleEntitiesByRSTag(tag, radius)
  return self:GetVisibleEntitiesByTypes(self.entity.relationship:GetTagTypes(tag), radius)
end

return VisionComponent

File: CSharpCodeBase/components/character/vision.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class vision :component{
 var VisionComponent = Class({__includes=Component, init = public void (self, entity){
       Component.init(self, entity);
       this.entity.vision = self;
       this.radius = this.entity.config:Get("visionRadius");
       this._tags = {}
 }})
 public void VisionComponent:IsTargetVisible(targetEntity){
   return targetEntity:GetEffectiveDistance(this.entity) < this.radius;
 }
 public void VisionComponent:GetVisibleEntities(radius){
   return RaycastUtils.GetEntitiesInRadius(this.entity:GetPosition(), radius  ||  this.radius);
 }
 public void VisionComponent:GetVisibleEntitiesByTypes(types, radius){
   return RaycastUtils.GetEntitiesInRadiusByTypes(this.entity:GetPosition(),;
     radius  ||  this.radius,;
     types);
 }
 public void VisionComponent:GetVisibleEntitiesByRSTag(tag, radius){
   return self:GetVisibleEntitiesByTypes(this.entity.relationship:GetTagTypes(tag), radius);
 }
 Component}}